src/core/utils/flash.cpp

```diff
@@ -0,0 +1,266 @@
+                uint16_t readLength = *aValueLength;
+
+                if (readLength > record.GetLength())
+                {
+                    readLength = record.GetLength();
+                }
+
+                otPlatFlashRead(&GetInstance(), mSwapIndex, offset + sizeof(record), aValue, readLength);
+            }
+
+            valueLength = record.GetLength();
+            error       = OT_ERROR_NONE;
+        }
+
+        index++;
+    }
+
+    if (aValueLength)
+    {
+        *aValueLength = valueLength;
+    }
+
+    return error;
+}
+
+otError Flash::Set(uint16_t aKey, const uint8_t *aValue, uint16_t aValueLength)
+{
+    return Add(aKey, true, aValue, aValueLength);
+}
+
+otError Flash::Add(uint16_t aKey, const uint8_t *aValue, uint16_t aValueLength)
+{
+    bool first = (Get(aKey, 0, NULL, NULL) == OT_ERROR_NOT_FOUND);
+
+    return Add(aKey, first, aValue, aValueLength);
+}
+
+otError Flash::Add(uint16_t aKey, bool aFirst, const uint8_t *aValue, uint16_t aValueLength)
+{
+    otError error = OT_ERROR_NONE;
+    Record  record;
+
+    record.Init(aKey, aFirst);
+    record.SetData(aValue, aValueLength);
+
+    if ((mSwapSize - record.GetSize()) < mSwapUsed)
+    {
+        Swap();
+        VerifyOrExit((mSwapSize - record.GetSize()) >= mSwapUsed, error = OT_ERROR_NO_BUFS);
+    }
+
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, record.GetSize());
+
+    record.SetAddCompleteFlag();
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, sizeof(RecordHeader));
+
+    mSwapUsed += record.GetSize();
+
+exit:
+    return error;
+}
+
+bool Flash::DoesValidRecordExist(uint32_t aOffset, uint16_t aKey) const
+{
+    RecordHeader record;
+    bool         rval = false;
+
+    for (; aOffset < mSwapUsed; aOffset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, aOffset, &record, sizeof(record));
+
+        if (record.IsValid() && record.IsFirst() && (record.GetKey() == aKey))
+        {
+            ExitNow(rval = true);
+        }
+    }
+
+exit:
+    return rval;
+}
+
+void Flash::Swap(void)
+{
+    uint8_t  dstIndex  = !mSwapIndex;
+    uint32_t dstOffset = kSwapMarkerSize;
+    Record   record;
+
+    otPlatFlashErase(&GetInstance(), dstIndex);
+
+    for (uint32_t srcOffset = kSwapMarkerSize; srcOffset < mSwapUsed; srcOffset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, sizeof(RecordHeader));
+
+        VerifyOrExit(record.IsAddBeginSet());
+
+        if (!record.IsValid() || DoesValidRecordExist(srcOffset + record.GetSize(), record.GetKey()))
+        {
+            continue;
+        }
+
+        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, record.GetSize());
+        otPlatFlashWrite(&GetInstance(), dstIndex, dstOffset, &record, record.GetSize());
+        dstOffset += record.GetSize();
+    }
+
+exit:
+    otPlatFlashWrite(&GetInstance(), dstIndex, 0, &sSwapActive, sizeof(sSwapActive));
```

Comment:
shall we store this in fixed little-endian?

Reply:
Is there a use case that concerns dealing with different architectures?

Reply:
None for now, but ARM supports to run on either big/little endian, I. prefer persist in fixed endian just for some unexpected cases. I don't have strong opinion on this.